### PR TITLE
Fix expectations in Deployment.

### PR DESCRIPTION
**Pull request by bgrant0607**
Working on a fix for #19299.

Probably still want eventual expiration, but if e2e doesn't pass without that, there's got to be a problem.

cc @janetkuo @kargakis @mqliang @nikhiljindal @ironcladlou

### CODE REVIEW

**Comment by bgrant0607** [comment]
All tests passed except for:
```
Deployment
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:72
  paused deployment should be ignored by the controller [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:62

  Feb 25 02:07:21.882: Err : nginx
  . Failed to remove deployment &{{ } {/api/v1/namespaces/e2e-tests-deployment-t46ni/pods 2875} [{{ } {nginx-cifwz-j9rn2 nginx-cifwz- e2e-tests-deployment-t46ni /api/v1/namespaces/e2e-tests-deployment-t46ni/pods/nginx-cifwz-j9rn2 7aa22eb7-dba7-11e5-a631-42010af00002 1511 %!s(int64=0) 2016-02-25 02:06:46 -0800 PST <nil> %!s(*int64=<nil>) map[name:nginx pod-template-hash:1900829565] map[kubernetes.io/created-by:{"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"ReplicaSet","namespace":"e2e-tests-deployment-t46ni","name":"nginx-cifwz","uid":"7a62371a-dba7-11e5-a631-42010af00002","apiVersion":"extensions","resourceVersion":"1158"}}
  ]} {[{default-token-919tm {%!s(*api.HostPathVolumeSource=<nil>) %!s(*api.EmptyDirVolumeSource=<nil>) %!s(*api.GCEPersistentDiskVolumeSource=<nil>) %!s(*api.AWSElasticBlockStoreVolumeSource=<nil>) %!s(*api.GitRepoVolumeSource=<nil>) %!s(*api.SecretVolumeSource=&{default-token-919tm}) %!s(*api.NFSVolumeSource=<nil>) %!s(*api.ISCSIVolumeSource=<nil>) %!s(*api.GlusterfsVolumeSource=<nil>) %!s(*api.PersistentVolumeClaimVolumeSource=<nil>) %!s(*api.RBDVolumeSource=<nil>) %!s(*api.FlexVolumeSource=<nil>) %!s(*api.CinderVolumeSource=<nil>) %!s(*api.CephFSVolumeSource=<nil>) %!s(*api.FlockerVolumeSource=<nil>) %!s(*api.DownwardAPIVolumeSource=<nil>) %!s(*api.FCVolumeSource=<nil>) %!s(*api.AzureFileVolumeSource=<nil>) %!s(*api.ConfigMapVolumeSource=<nil>)}}] [{nginx nginx [] []  [] [] {map[] map[]} [{default-token-919tm %!s(bool=true) /var/run/secrets/kubernetes.io/serviceaccount}] %!s(*api.Probe=<nil>) %!s(*api.Probe=<nil>) %!s(*api.Lifecycle=<nil>) /dev/termination-log Always %!s(*api.SecurityContext=<nil>) %!s(bool=false) %!s(bool=false) %!s(bool=false)}] Always %!s(*int64=0xc2084f6128) %!s(*int64=<nil>) ClusterFirst map[] default e2e-gce-master-0-minion-6zpd %!s(*api.PodSecurityContext=&{false false false <nil> <nil> <nil> [] <nil>}) []} {Running [{Ready True 0001-01-01 00:00:00 +0000 UTC 2016-02-25 02:06:51 -0800 PST  }]   10.240.0.6 10.245.4.11 2016-02-25 02:06:47 -0800 PST [{nginx {%!s(*api.ContainerStateWaiting=<nil>) %!s(*api.ContainerStateRunning=&{{{63591991609 0 0x2cb8520}}}) %!s(*api.ContainerStateTerminated=<nil>)} {%!s(*api.ContainerStateWaiting=<nil>) %!s(*api.ContainerStateRunning=<nil>) %!s(*api.ContainerStateTerminated=<nil>)} %!s(bool=true) %!s(int=0) nginx docker://ae8e1e9c54b3fcaed87ff6ded395690df8dbd2bb4e9842485ab36a89d491c6f9 docker://1d36039453eb391f71cb86cc08b7107b9105063734c23317299e8679c7d98f16}]}}]} pods : %!v(MISSING)

  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:197
```

**Comment by bgrant0607** [comment]
The paused test fails to delete all of the Deployment's pods at the very end.

**Comment by bgrant0607** [comment]
Ok. I changed Deployment to do straightforward refcounting of pod additions and deletions. This made all deployment tests pass for me, except for pause.

I need someone else (or multiple someones) to look at pause.

And I would like someone to change the ExpirationCache to permit explicit timeout instead of expiration on Get. The timestamp should be reset every time syncDeployment is able to make progress (i.e., stores are synced and expectations are met) -- this should be when SetExpectations is called.

Without vectors of logical clocks, this is not going to be bulletproof, but it's at least not blatantly wrong.

File: pkg/controller/deployment/deployment_controller.go [hunk 1]
```diff
@@ -1081,13 +1101,29 @@ func (dc *DeploymentController) scaleReplicaSetAndRecordEvent(rs *extensions.Rep
...
+		return false, nil, fmt.Errorf("couldn't get key for deployment %#v: %v", deployment, err)
+	}
+	var scalingOperation string
+	// Set expectations first, because if the update is successful, the expectations will be handled asynchronously immediately.
 	if rs.Spec.Replicas < newScale {
 		scalingOperation = "up"
+		dc.podExpectations.RaiseExpectations(dKey, newScale-rs.Spec.Replicas, 0)
+	} else {
+		scalingOperation = "down"
+		dc.podExpectations.LowerExpectations(dKey, 0, rs.Spec.Replicas-newScale)
```
Comment (madhusudancs) [inline pkg/controller/deployment/deployment_controller.go]: Shouldn't this be:

```go
dc.podExpectations.RaiseExpectations(dKey, 0, rs.Spec.Replicas-newScale)
```

or

```go
dc.podExpectations.LowerExpectations(dKey, 0, newScale-rs.Spec.Replicas)
```

Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: Good catch.

**Comment by madhusudancs** [comment]
LGTM, other than one minor comment. Please feel free to apply the label after making that change.

**Comment by bgrant0607** [comment]
Addressed feedback and did some more cleanup (conversion of Set/Delete to Raise/Lower). Enabled flaky tests in PR builder.

File: pkg/controller/controller_utils.go [hunk 2]
```diff
@@ -188,10 +199,10 @@ type ControlleeExpectations struct {
 	key string
 }
 
-// Seen decrements the add and del counters.
-func (e *ControlleeExpectations) Seen(add, del int64) {
-	atomic.AddInt64(&e.add, -add)
-	atomic.AddInt64(&e.del, -del)
+// Seen increments the add and del counters.
```
Comment (janetkuo) [inline pkg/controller/controller_utils.go]: change the comment to `Add`

**Comment by janetkuo** [comment]
Tests failed due to timeout:
```
Kubernetes e2e suite.Deployment RollingUpdateDeployment should scale up and down in the right order
Kubernetes e2e suite.Deployment RecreateDeployment should delete old pods and create new ones
Kubernetes e2e suite.Deployment deployment should delete old replica sets
Kubernetes e2e suite.Deployment RollingUpdateDeployment should delete old pods and create new ones
Kubernetes e2e suite.Deployment deployment should support rollover
```
Can't find the controller manger log in the bucket.

**Comment by bgrant0607** [comment]
@janetkuo Thanks. I'll take a look. Those are all the same tests that have been failing, so looks like there's still a problem.

**Comment by janetkuo** [comment]
Tests seem to be still failing and the kube-controller-manager.log is huge (13G) so I didn't try to open it. Tailing the file all I see is this repeating error message:
```
E0225 14:50:09.087732  124308 deployment_controller.go:408] Error syncing deployment e2e-tests-deployment-hwi9u/redis-deployment: pod expectations not met yet for e2e-tests-deployment-hwi9u/redis-deployment in syncDeployment
```

**Comment by bgrant0607** [comment]
Added expiration back so I can (hopefully) see what expectations aren't being satisfied.

**Comment by bgrant0607** [comment]
Controller manager log was huge because I was always re-enqueueing on failure, only to fail again.

Pods from ReplicaSets created with non-zero replicas weren't accounted for.

It also appears that not all expected deletions happen.

Could be faulty error handling.
```
I0226 00:35:40.222253       6 controller_utils.go:188] Raising expectations &{add:0 del:3 key:e2e-tests-deployment-y6wmn/redis-deployment-3 timestamp:{sec:63592043740 nsec:81518825 loc:0x263dd00}}
I0226 00:35:40.224626       6 controller_utils.go:179] Lowering expectations &{add:0 del:3 key:e2e-tests-deployment-y6wmn/redis-deployment-3 timestamp:{sec:63592043740 nsec:81518825 loc:0x263dd00}}
I0226 00:35:40.224663       6 deployment_controller.go:419] Finished syncing deployment "e2e-tests-deployment-y6wmn/redis-deployment-3" (143.246792ms)
E0226 00:35:40.224690       6 deployment_controller.go:408] Error syncing deployment e2e-tests-deployment-y6wmn/redis-deployment-3: replicasets "nginx-controller" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
I0226 00:35:40.224749       6 controller_utils.go:133] Controller still waiting on expectations &controller.ControlleeExpectations{add:0, del:3, key:"e2e-tests-deployment-y6wmn/redis-deployment-3", timestamp:time.Time{sec:63592043740, nsec:81518825, loc:(*time.Location)(0x263dd00)}}
I0226 00:35:40.224789       6 deployment_controller.go:419] Finished syncing deployment "e2e-tests-deployment-y6wmn/redis-deployment-3" (52.129µs)
```

**Comment by bgrant0607** [comment]
It appears to be pretty common for deletions to be observed after expectations are determined to be satisfied and are reset.

**Comment by bgrant0607** [comment]
Something fishy is going on with deletes, but I don't have time to investigate more. This seems to make all the tests pass.

PTAL

**Comment by bgrant0607** [comment]
BTW, this also makes the controller manager log comparable in size to the other components.

File: pkg/controller/controller_utils.go [hunk 3]
```diff
@@ -183,15 +206,16 @@ type Expectations interface {
...
+	del       int64
+	key       string
+	timestamp time.Time
 }
 
-// Seen decrements the add and del counters.
-func (e *ControlleeExpectations) Seen(add, del int64) {
-	atomic.AddInt64(&e.add, -add)
-	atomic.AddInt64(&e.del, -del)
+// Add increments the add and del counters.
```
Comment (madhusudancs) [inline pkg/controller/controller_utils.go]: This is much more readable and easier to comprehend than the `Seen()` language. Thanks!

**Comment by madhusudancs** [comment]
LGTM. I will let @janetkuo to add the label after taking a look.

**Comment by kargakis** [comment]
LGTM

**Comment by bgrant0607** [comment]
Merging to fix tests.

**Pull request merged**